draft-ietf-httpbis-http2-secondary-certs.md

```diff
@@ -753,6 +753,80 @@ any stream which references the failing certificate in question or process the
 requests as unauthenticated and provide error information at the HTTP semantic
 layer.
 
+# Certificate Extensions {#extension}
+
+## Allowed Primary Domain Certificate Extension {#extension-primary}
+
+The Allowed Primary Domain extension allows certificates to limit their use with
+Secondary Certificate Authentication.  The identities listed in this extension
+are restrictions asserted by the requester of the certificate and are not
+verified by the CA.  Conforming CAs SHOULD mark the allowedPrimaryDomain
+extension as non-critical.
+
+Each allowed primary domain is represented as an IA5String. The name MUST be in
+the "preferred name syntax", as specified by Section 3.5 of {{!RFC1034}} and as
+modified by Section 2.1 of {{!RFC1123}}.  Note that while uppercase and
+lowercase letters are allowed in domain names, no significance is attached to
+the case. In addition, while the string " " is a legal domain name,
+affiliatedDomain extensions MUST NOT contain " ".  Rules for encoding
+internationalized domain names are specified in Section 7.2 of {{!RFC5280}}.
+
+If the affiliatedDomain extension is present, the sequence MUST contain at least
+one entry.  Unlike the subject field, conforming CAs MUST NOT issue certificates
+with an affiliatedDomain extension containing empty GeneralName fields.  Clients
+that encounter such a certificate when processing a certification path MUST
+ignore that extension.
+
+The wildcard character "*" MAY be used to represent that any set of subdomains
+is acceptable.  This character MUST be the entirety of the first label if used,
+and MUST NOT occur elsewhere.  (That is, "*.example.com" and "*" are acceptable,
+but "www.*.com" and "w*.example.com" are not).
+
+    id-ce-allowedPrimaryDomain OBJECT IDENTIFIER ::=  { id-ce TBD1 }
+
+    AllowedPrimaryDomain ::= SEQUENCE SIZE (1..MAX) OF IA5String
+
+## Affiliated Domain Certificate Extension {#extension-affiliated}
+
+The Affiliated Domain extension allows groups of certificates under the same
+control to be identified without giving each certificate full identity with the
+domain.  The identities stated as affiliated domains are not bound to the
+subject of the certificate, but this extension indicates that there is a common
+management relationship between the subject and the affiliated domains.
```

Comment:
Here I think that the list of names is a genuine problem.  For this, I would recommend instead that this include a single name.  Keep the requirement that it be a domain name under the control of the entity that is serving these connections, but don't constrain it.  Basically, say that it must be validated as though it were subjectAltName (see below for more on this point.)

The problem with multiple values is that you can have a primary with A and B, a secondary with B and C, and a third secondary with C and D and that might be allowed, even though the third secondary shares no common points with the primary.

Similarly, you might have a secondary with A and D, which means that you can have secondaries with no common affiliated domains.

The reason you validate this is so that an attacker who compromises a key for foo.example.com and can cause a connection to be established to example.attacker can't get the two together because example.attacker because both certificates no longer match up.

Reply:
Yes, but I think the chaining is necessary (and that's why the affiliated and the allowed are separate), or at least strongly desirable.  Consider:

- Primary:  (Affiliated:  M)
- First Secondary:  (Affiliated:  A, Allowed:  *)
- Second Secondary:  (Allowed:  A, Affiliated:  Z)
- Third Secondary:  (Allowed:  A, Affiliated: Y)

It's fine to have secondaries without overlapping affiliations, provided that there's a clear chain in the allowed path.

Reply:
OK, so I *really* don't want chain building here.  But I can see how you get there and it sucks.  Need to think more about this.